torch/csrc/distributed/rpc/rref_context.cpp

```diff
@@ -360,12 +405,15 @@ void RRefContext::addPendingChild(
 }
 
 void RRefContext::delPendingChild(const ForkId& forkId) {
-  std::lock_guard<std::mutex> lock(mutex_);
-  auto iter = pendingChildren_.find(forkId);
-  TORCH_INTERNAL_ASSERT(
-      iter != pendingChildren_.end(),
-      "Inconsistent states: attempt to delete a non-exist child fork.");
-  pendingChildren_.erase(iter);
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    auto iter = pendingChildren_.find(forkId);
+    TORCH_INTERNAL_ASSERT(
+        iter != pendingChildren_.end(),
+        "Inconsistent states: attempt to delete a non-exist child fork.");
+    pendingChildren_.erase(iter);
```

Comment:
cannot directly erase the `UserRRef` here, because that could trigger `delUser` which would then acquire lock on `mutex_` to modify `confirmedUsers_`. As the lock is already acquired in this block, it would deadlock. 

#34610 solved this problem by creating a temporary `c10::intrusive_ptr<RRef> deletedUser` to keep the deleted UserRRef alive in this block.

Reply:
Thanks for explaining the subtlety here. Updating with comments.